src/Juvix/Compiler/Pipeline/Loader/PathResolver.hs

```diff
@@ -401,33 +401,74 @@ expectedPath' m = do
   return PathInfoTopModule {..}
 
-re ::
-  forall r a.
-  (Members '[TaggedLock, Reader EntryPoint, Files, Error JuvixError, Error DependencyError, GitClone, EvalFileEff] r) =>
+runPathResolver2 ::
+  forall r a v.
+  (v ~ '[TaggedLock, Reader EntryPoint, Files, Error JuvixError, Error DependencyError, GitClone, EvalFileEff], Members v r) =>
+  ResolverState ->
+  ResolverEnv ->
   Sem (PathResolver ': r) a ->
-  Sem (Reader ResolverEnv ': State ResolverState ': r) a
-re = reinterpret2H helper
+  Sem r (ResolverState, a)
+runPathResolver2 st topEnv arg = do
+  ( reinterpretHCommon2
+      ( runState st
+          . runReader topEnv
+      )
+      handler
+    )
+    arg
   where
-    helper ::
-      forall rInitial x.
-      PathResolver (Sem rInitial) x ->
-      Tactical PathResolver (Sem rInitial) (Reader ResolverEnv ': (State ResolverState ': r)) x
-    helper = \case
-      RegisterDependencies forceUpdateLockfile -> registerDependencies' forceUpdateLockfile >>= pureT
-      ExpectedPathInfoTopModule m -> expectedPath' m >>= pureT
-      WithPath m a -> do
-        x :: Either PathResolverError (Path Abs Dir, Path Rel File) <- resolvePath' m
-        oldroot <- asks (^. envRoot)
-        x' <- pureT x
-        a' <- bindT a
-        st' <- get
-        let root' = case x of
-              Left {} -> oldroot
-              Right (r, _) -> r
-        raise (evalPathResolver' st' root' (a' x'))
-
-evalPathResolver' :: (Members '[TaggedLock, Reader EntryPoint, Files, Error JuvixError, Error DependencyError, GitClone, EvalFileEff] r) => ResolverState -> Path Abs Dir -> Sem (PathResolver ': r) a -> Sem r a
-evalPathResolver' st root = fmap snd . runPathResolver' st root
+    handler ::
+      forall t.
+      (Members v t) =>
+      EffectHandler PathResolver (Reader ResolverEnv ': State ResolverState ': t)
+    handler localEnv = \case
+      RegisterDependencies forceUpdateLockfile -> registerDependencies' forceUpdateLockfile
+      ExpectedPathInfoTopModule m -> expectedPath' m
+      WithPath
+        m
+        ( a ::
+            Either PathResolverError (Path Abs Dir, Path Rel File) ->
+            Sem localEs x
+          ) -> do
+          x :: Either PathResolverError (Path Abs Dir, Path Rel File) <- resolvePath' m
+          let y :: Sem localEs x = a x
+          oldroot <- asks (^. envRoot)
+          st' <- get
+          let root' = case x of
+                Left {} -> oldroot
+                Right (r, _) -> r
+          e <- ask
+          let _envSingleFile :: Maybe (Path Abs File)
+              _envSingleFile
+                | e ^. entryPointPackageType == GlobalStdlib = e ^. entryPointModulePath
+                | otherwise = Nothing
+              env' :: ResolverEnv
+              env' =
+                ResolverEnv
+                  { _envRoot = root',
+                    _envLockfileInfo = Nothing,
+                    _envSingleFile
+                  }
+
+          let runner ::
```

Comment:
It looks to me like you're unnecessarily jumping through tons of hoops here.

You constructed `st'` and `env'`, so IIUC simply calling `local (const env') $ bracket (put st') (put st) $ unlift y` instead of, well, all of this here should do the trick.

Reply:
thank you! we've simplified the code following your advice